salt/metalk8s/internal/sreport/files/metalk8s.py

```diff
@@ -0,0 +1,123 @@
+from sos.plugins import Plugin, RedHatPlugin
+from os import path
+
+
+class metalk8s(Plugin, RedHatPlugin):
+
+    """Metalk8s plugin
+    """
+
+    plugin_name = 'metalk8s'
+    packages = ('kubernetes', 'kubernetes-master')
+    profiles = ('container',)
+    files = ("/etc/kubernetes/admin.conf",)
+
+    option_list = [
+        ("all", "also collect all namespaces output separately",
+            'slow', False),
+        ("describe", "capture descriptions of all kube resources",
+            'fast', False),
+        ("podlogs", "capture logs for pods", 'slow', False),
+    ]
+
+    def check_is_master(self):
+        if any([
+            path.exists("/etc/kubernetes/admin.conf")
+        ]):
+            return True
+        return False
+
+    def setup(self):
+        self.add_copy_spec("/etc/kubernetes")
+        self.add_copy_spec("/var/log/pods")
+        self.add_copy_spec("/var/log/metalk8s-bootstrap.log")
+
+        svcs = [
+            'kubelet',
+        ]
+
+        for svc in svcs:
+            self.add_journal(units=svc)
+
+        # We can only grab kubectl output from the master
+        if self.check_is_master():
+            kube_cmd = "kubectl "
+            if path.exists('/etc/kubernetes/admin.conf'):
+                kube_cmd += "--kubeconfig=/etc/kubernetes/admin.conf"
+
+            kube_get_cmd = "get -o json "
+            for subcmd in ['version', 'config view']:
+                self.add_cmd_output('%s %s' % (kube_cmd, subcmd))
+
+            # get all namespaces in use
+            kn = self.get_command_output('%s get namespaces' % kube_cmd)
+            knsps = [n.split()[0] for n in kn['output'].splitlines()[1:] if n]
+
+            resources = [
+                'pods',
+                'deploy',
+                'rc',
+                'services',
+                'ds',
+                'cm',
+            ]
+
+            # nodes and pvs are not namespaced, must pull separately.
+            # Also collect master metrics
+            self.add_cmd_output([
+                "{} get -o json nodes".format(kube_cmd),
+                "{} get -o json pv".format(kube_cmd),
+                "{} get --raw /metrics".format(kube_cmd)
+            ])
+
+            for n in knsps:
+                knsp = '--namespace=%s' % n
+                if self.get_option('all'):
+                    k_cmd = '%s %s %s' % (kube_cmd, kube_get_cmd, knsp)
```

Comment:
`k_cmd`, why the `k_` prefix? What does it means?

If you can't use `cmd` because you already have variable named like that, maybe it's the sign that we should split this function (it becames too big).

Reply:
I agree this is not the best code you can read but this is half vendored/modified code and I fear it become hard to adapt to the upstream if we derive so much

Reply:
Hum, not very fond of the half vendored/modified…

- Either it's vendored code, and we should keep it as-is, with no modification.
- Either we take the ownership and we can rewrite it as we want.

Falling between two stools gives us the worse of both world:
- you don't want to fix the code, because it's vendored
- but you've still modified it enough to be pain when we will want to follow upstream (conflicts).

Reply:
make sens, going to modify it, still it is ready for a functional review